trace_replay/block_cache_tracer.cc

```diff
@@ -0,0 +1,223 @@
+
+#include "db/db_impl/db_impl.h"
+#include "rocksdb/slice.h"
+#include "util/coding.h"
+#include "util/hash.h"
+#include "util/string_util.h"
+
+namespace rocksdb {
+
+namespace {
+const unsigned int kCharSize = 1;
+bool ShouldTraceReferencedKey(const BlockCacheTraceRecord& record) {
+  return (record.block_type == TraceType::kBlockTraceDataBlock) &&
+         (record.caller == BlockCacheLookupCaller::kUserGet ||
+          record.caller == BlockCacheLookupCaller::kUserMGet);
+}
+}  // namespace
+
+BlockCacheTraceWriter::BlockCacheTraceWriter(
+    Env* env, const TraceOptions& trace_options,
+    std::unique_ptr<TraceWriter>&& trace_writer)
+    : env_(env),
+      trace_options_(trace_options),
+      trace_writer_(std::move(trace_writer)) {}
+
+bool BlockCacheTraceWriter::ShouldTrace(const BlockCacheTraceRecord& record) {
+  if (trace_options_.sampling_frequency == 0 ||
+      trace_options_.sampling_frequency == 1) {
+    return true;
+  }
+  // We use spatial downsampling so that we have a complete access history for a
+  // block.
+  const uint64_t hash = GetSliceNPHash64(Slice(record.block_key));
+  return hash % trace_options_.sampling_frequency == 0;
+}
+
+Status BlockCacheTraceWriter::WriteBlockAccess(
+    const BlockCacheTraceRecord& record) {
+  uint64_t trace_file_size = trace_writer_->GetFileSize();
+  if (trace_file_size > trace_options_.max_trace_file_size) {
+    return Status::OK();
+  }
+  if (!ShouldTrace(record)) {
+    return Status::OK();
+  }
+  Trace trace;
+  trace.ts = record.access_timestamp;
+  trace.type = record.block_type;
+  PutLengthPrefixedSlice(&trace.payload, record.block_key);
+  PutFixed64(&trace.payload, record.block_size);
+  PutFixed32(&trace.payload, record.cf_id);
+  PutLengthPrefixedSlice(&trace.payload, record.cf_name);
+  PutFixed32(&trace.payload, record.level);
+  PutFixed32(&trace.payload, record.sst_fd_number);
+  trace.payload.push_back(record.caller);
+  trace.payload.push_back(record.is_cache_hit);
+  trace.payload.push_back(record.no_insert);
+  if (ShouldTraceReferencedKey(record)) {
+    PutLengthPrefixedSlice(&trace.payload, record.referenced_key);
+    PutFixed64(&trace.payload, record.num_keys_in_block);
+    trace.payload.push_back(record.is_referenced_key_exist_in_block);
+  }
+  std::string encoded_trace;
+  TracerHelper::EncodeTrace(trace, &encoded_trace);
+  trace_writer_mutex_.Lock();
+  const Status s = trace_writer_->Write(encoded_trace);
+  trace_writer_mutex_.Unlock();
+  return s;
+}
+
+Status BlockCacheTraceWriter::WriteHeader() {
+  Trace trace;
+  trace.ts = env_->NowMicros();
+  trace.type = TraceType::kTraceBegin;
+  PutLengthPrefixedSlice(&trace.payload, kTraceMagic);
+  PutFixed32(&trace.payload, kMajorVersion);
+  PutFixed32(&trace.payload, kMinorVersion);
+  std::string encoded_trace;
+  TracerHelper::EncodeTrace(trace, &encoded_trace);
+  trace_writer_mutex_.Lock();
+  const Status s = trace_writer_->Write(encoded_trace);
```

Comment:
Why `const`?

Reply:
I removed this line since we are using InstrumentedMutexLock now.